### PR TITLE
Advanced Handling of Dependency Issues

### PR DESCRIPTION
**Pull request by yxususe**
The approach for ticket https://progress.opensuse.org/issues/9498 Advanced Handling of Dependency issues

- needles [gl#os-autoinst-needles-sles#435](https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/435)

### CODE REVIEW

File: lib/y2logsstep.pm [hunk 1]
```diff
@@ -107,6 +107,81 @@ sub check_and_record_dependency_problems {
...
+            wait_screen_change {
+                send_key 'spc';                            # select it
+            }
+            send_key 'alt-o';                              # OK
+        } while (check_screen 'dependancy-issue', 5);
+
+        wait_screen_change {
+            send_key 'alt-a';                              # Accept
+        }
+        send_key 'alt-a', 3;                               # Accept
```
Comment (okurz) [inline lib/y2logsstep.pm]: Please avoid the second argument on send_key. It's a boolean value (so only 0 or 1 should be used) but internally it calls wait_idle which we like to avoid

File: lib/y2logsstep.pm [hunk 2]
```diff
@@ -107,6 +107,81 @@ sub check_and_record_dependency_problems {
...
+            }
+            wait_screen_change {
+                send_key 'spc';                            # select it
+            }
+            send_key 'alt-o';                              # OK
+        } while (check_screen 'dependancy-issue', 5);
+
+        wait_screen_change {
+            send_key 'alt-a';                              # Accept
+        }
```
Comment (okurz) [inline lib/y2logsstep.pm]: to write it a bit more concise I recommend

```perl
wait_screen_change { send_key 'alt-a' };
```

File: tests/installation/select_patterns_and_packages.pm [hunk 3]
```diff
@@ -69,6 +69,10 @@ sub gotopatterns {
         $self->record_dependency_issues;
     }
 
+    if (check_screen('dependancy-issue', 10) && get_var("BREAK_DEPS")) {
```
Comment (okurz) [inline tests/installation/select_patterns_and_packages.pm]: try to avoid timeout values which are neither 0 nor bigger than the default because they tend to be not robust, e.g. when the machine is a bit loaded. It should work with argument 0, especially when combined with synchronization methods used in before, e.g. `wait_screen_change` or `wait_still_screen`

This comment also applies to all other check_screens you introduced

**Comment by yxususe** [comment]
I'll test it under scenario Development SLE15.

File: tests/installation/select_patterns_and_packages.pm [hunk 4]
```diff
@@ -65,8 +65,11 @@ sub gotopatterns {
         send_key 'ret';
     }
 
-    if (check_screen('dependancy-issue', 10) && get_var("WORKAROUND_DEPS")) {
-        $self->record_dependency_issues;
+    if (check_screen('dependancy-issue') && get_var("WORKAROUND_DEPS")) {
+        $self->workaround_dependency_issues;
+    }
+    if (check_screen('dependancy-issue') && get_var("BREAK_DEPS")) {
```
Comment (okurz) [inline tests/installation/select_patterns_and_packages.pm]: Better avoid check_screen with a timeout of default or lower but use multi-tag assert_screen in combination with match_has_tag

**Comment by yxususe** [comment]
WIP updated. 
A successfully finished testrun with soft failure can be found here: sle-15-Leanos-DVD-x86_64 http://f146.suse.de/tests/1051125. (with parameter BREAK_DEPS=1)
I'm still testing it for migration_offline and package-dependency.

**Comment by okurz** [comment]
How does that differ from https://openqa.suse.de/tests/1051125 ?

**Comment by yxususe** [comment]
The testrun 1051125 doesn't use paramter BREAK_DEPS=1 but WORKAROUND=1.

**Comment by yxususe** [comment]
Dependency issues can be handled now with paramter BREAK_DEPS=1:
http://f146.suse.de/tests/678   for scenario migration_offline_sled11_sp4_allpatterns_sdk
http://f146.suse.de/tests/679   for scenario sle-15-Leanos-btrfs
http://f146.suse.de/tests/671   for scenario package-dependency

For package-dependency I commented out the failing part for now, because it seems to be able to handle the error (some packages need to be selected) by itself. However I'll still take a look.

I have spent a lot of time testing the difference between assert_screen with combination of match_has_tag and check_screen. And I chose check_screen for higher reliability.

**Comment by yxususe** [comment]
I'm now testing zdup and change_desktop.

File: lib/y2logsstep.pm [hunk 5]
```diff
@@ -57,61 +58,130 @@ sub get_to_console() {
...
-            }
-        };
-        wait_screen_change {
+# to workaround dependency issues
+sub workaround_dependency_issues {
+    if (check_var('VIDEOMODE', 'text')) {
+        while (check_screen('dependency-issue-text', 5)) {
+            send_key 'alt-s';
+            send_key 'ret';
+            send_key 'alt-o';
```
Comment (okurz) [inline lib/y2logsstep.pm]: I assume this unsynchronized sequence of keys is unreliable, better guard with `wait_screen_change` around for the ones before the next synchronisation point, e.g.

```perl
wait_screen_change { send_key 'alt-s' };
wait_screen_change { send_key 'ret' };
send_key 'alt-o';
wait_still_screen 1;
```

File: lib/y2logsstep.pm [hunk 6]
```diff
@@ -57,61 +58,130 @@ sub get_to_console() {
...
+        }
+    }
+    else {
+        while (check_screen('dependency-issue', 5)) {
+            send_key 'alt-2';                                 # 2 is the option to break dependency
+            send_key 'spc';                                   # select it
+            send_key 'alt-o';                                 # OK
+            wait_still_screen 1;
+        }
+    }
```
Comment (okurz) [inline lib/y2logsstep.pm]: IMHO that would be way more readable when you turn the if+while around. Let me try without promising that I did not change behaviour:

```perl
while (check_screen('dependency-issue', 5)) {
    if (check_var('VIDEOMODE', 'text')) {
        wait_screen_change { send_key 'alt-s' };
        send_key 'down' while check_screen('break-not-highlighted', 5);
        wait_screen_change { send_key 'ret' }
    }
    else {
        wait_screen_change { send_key 'alt-2' };
        wait_screen_change { send_key 'spc' };
    }
    send_key $cmd{ok};
    wait_still_screen 1;
}
```

File: lib/y2logsstep.pm [hunk 7]
```diff
@@ -57,61 +58,130 @@ sub get_to_console() {
...
-        wait_screen_change {
-            send_key 'alt-a';
-        };
+        $self->workaround_dependency_issues;
+    }
+    elsif (get_var 'BREAK_DEPS') {
+        $self->break_dependency;
+    }
+    else {
+        die 'Dependency Problems';
```
Comment (okurz) [inline lib/y2logsstep.pm]: s/Problems/problems/

File: lib/y2logsstep.pm [hunk 8]
```diff
@@ -57,61 +58,130 @@ sub get_to_console() {
...
+    }
+    if (check_screen('dependency-issue-fixed', 2)) {
+        send_key 'alt-a';    # Accept
+    }
+    wait_still_screen 2;
+
+  DO_CHECKS:
+    while (check_screen('accept-licence', 2)) {
```
Comment (okurz) [inline lib/y2logsstep.pm]: I am not convinced that the `check_screen` approach is as stable as the multi-tag assert. Just image a SUT being slightly more busy than usual and it would take more than 2 seconds to show up that dialog. You would fail. I don't agree with this change here. I need better reasons to be convinced that an approach with check_screen is "more reliable" as you state

File: tests/installation/select_patterns_and_packages.pm [hunk 9]
```diff
@@ -27,7 +27,7 @@ my $secondrun = 0;    # bsc#1029660
 
 sub accept3rdparty {
     #Third party licenses sometimes appear
-    while (check_screen([qw(3rdpartylicense automatic-changes inst-overview)], 15)) {
+    while (check_screen([qw(3rdpartylicense automatic-changes inst-overview)], 5)) {
```
Comment (okurz) [inline tests/installation/select_patterns_and_packages.pm]: looks like an unrelated change

File: lib/y2logsstep.pm (hunk 5 above)
Comment (yxususe) [inline lib/y2logsstep.pm]: OK. Thanks for the suggestion.

File: lib/y2logsstep.pm (hunk 6 above)
Comment (yxususe) [inline lib/y2logsstep.pm]: Yes, however my attention was to reduce the time of judgement, which is, I judge only once if it is textmode or not.

File: lib/y2logsstep.pm (hunk 7 above)
Comment (yxususe) [inline lib/y2logsstep.pm]: OK.

File: tests/installation/select_patterns_and_packages.pm (hunk 9 above)
Comment (yxususe) [inline tests/installation/select_patterns_and_packages.pm]: OK. I can reverse it.

**Comment by yxususe** [comment]
Test of zdup and change_desktop was done. However it didn't really meet dependency issues.

File: lib/y2logsstep.pm (hunk 6 above)
Comment (yxususe) [inline lib/y2logsstep.pm]: wait_screen_change { send_key 'alt-2' }; will cause an endless loop, because the screen almost doesn't change at all. It is just a mouse click in the option 2.

Comment (yxususe) [inline lib/y2logsstep.pm]: wait_screen_change { send_key 'spc' }; also causes an endless loop, because the screen doesn't change, it is to focus the mouse on the option "OK - Try Again".

File: lib/y2logsstep.pm (hunk 8 above)
Comment (yxususe) [inline lib/y2logsstep.pm]: Here after changing into while (match_has_tag 'accept-licence') (others similar) it is not reliable any more. The test ends up doing one thing: send_key 'alt-a'; 
Testruns can be viewed here:
http://f146.suse.de/tests/701
http://f146.suse.de/tests/702

**Comment by yxususe** [comment]
The modification to files are necessary and can't be separated any more because of the renaming of functions' name. I'll test the scenarios with parameter workaround, and see if it still works.

**Comment by sysrich** [comment]
Thanks - can you also fix the commit messages? we currently have 8 commits all with a commit message of 'WIP' - we need to know what each commit is attempting to do, not only for reviewing but if we ever need to go back and debug the test in the future.

**Comment by yxususe** [comment]
Yes, somehow git rebase didn't work, I try to fix it.

**Comment by yxususe** [comment]
`wait_screen_change { send_key 'alt-o' };` also makes the test endless waiting because the dependency issue popup window looks quite similar (after one dependency issue is resolved, the screen seems not to be changed to openqa). See http://f146.suse.de/tests/720

File: lib/y2logsstep.pm (hunk 5 above)
Comment (yxususe) [inline lib/y2logsstep.pm]: Most of the time the screen almost doesn't change at all, thus go back to `send_key`.

**Comment by yxususe** [comment]
geekotest@shaka:~> /usr/share/openqa/script/clone_job.pl --from https://openqa.suse.de 1057906 WORKAROUND_DEPS=1
Created job #721: sle-12-SP3-Desktop-DVD-x86_64-Build0291-migration_offline_sled11_sp4_allpatterns_sdk@64bit

geekotest@shaka:~> /usr/share/openqa/script/clone_job.pl --from https://openqa.suse.de 1057906 BREAK_DEPS=1
Created job #722: sle-12-SP3-Desktop-DVD-x86_64-Build0291-migration_offline_sled11_sp4_allpatterns_sdk@64bit

geekotest@shaka:~> /usr/share/openqa/script/clone_job.pl --from https://openqa.suse.de 1057906
downloading
Created job #723: sle-12-SP3-Desktop-DVD-x86_64-Build0291-migration_offline_sled11_sp4_allpatterns_sdk@64bit

For scenario migration_offline_sled11_sp4_allpatterns_sdk,
http://f146.suse.de/tests/721 is the testrun with parameter WORKAROUND_DEPS=1
http://f146.suse.de/tests/722 is the testrun with parameter BREAK_DEPS=1
http://f146.suse.de/tests/723 is the testrun with no paramter.

**Comment by yxususe** [comment]
geekotest@shaka:~> /usr/share/openqa/script/clone_job.pl --from https://openqa.suse.de 1051125
Created job #735: sle-15-Leanos-DVD-x86_64-Build42.1-btrfs@64bit

For the scenario sle-15-Leanos-btrfs,
http://f146.suse.de/tests/733 is the testrun with parameter BREAK_DEPS=1 
http://f146.suse.de/tests/734 is the testrun with parameter WORKAROUND_DEPS=1
http://f146.suse.de/tests/735 is the testrun with no parameter

**Comment by yxususe** [comment]
For the scenario sle-15-Leanos-btrfs,
geekotest@shaka:~> /usr/share/openqa/script/clone_job.pl --from http://openqa.suse.de 1051125 BREAK_DEPS=1
geekotest@shaka:~> /usr/share/openqa/script/clone_job.pl --from http://openqa.suse.de 1051125 WORKAROUND_DEPS=1
geekotest@shaka:~> /usr/share/openqa/script/clone_job.pl --from http://openqa.suse.de 1051125
The three commands makes no difference. Why are the parameters not accepted? What am I missing? For scenario migration_offline_sled11_sp4_allpatterns_sdk, it worked fine.

**Comment by yxususe** [comment]
OK, till now, what I can say is, if BREAK_DEPS=1 is chosen, then the installation can be finished, however often the system won't be able to boot. If WORKAROUND_DEPS=1 is chosen, under some situations the system can't finish installation, sometimes it can.

**Comment by yxususe** [comment]
Now I'll test it with VIDEOMODE=text.

**Comment by yxususe** [comment]
PR updated. For dependency issue under VIDEOMODE=text I'll do a new PR.

**Comment by yxususe** [comment]
Corresponding PR of needles: 
https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/435?diff_id=8783#a1f13b3bc20a296e08c212be9c56c706c10abc4f_0_18
